runtime/lua/vim/treesitter/query.lua

```diff
@@ -899,16 +915,35 @@ function Query:iter_captures(node, source, start, stop)
       return
     end
 
-    if not match_preds(self, match, source) then
-      local match_id = match:info()
-      cursor:remove_match(match_id)
-      if end_line and captured_node:range() > end_line then
-        return nil, captured_node, nil, nil
-      end
-      return iter(end_line) -- tail call: try next match
+    local match_id, pattern = match:info()
+
+    --- @type vim.treesitter.query.TSMetadata
+    local metadata
+    if match_id <= highest_cached_match_id then
+      metadata = match_cache[match_id]
     end
 
-    local metadata = apply_directives(self, match, source)
+    if not metadata then
+      local patterns = self.processed_patterns[pattern]
+      if patterns then
+        local captures = match:captures()
+
+        if not self:match_preds(patterns.preds, pattern, captures, source) then
+          cursor:remove_match(match_id)
+          if end_line and captured_node:range() > end_line then
+            return nil, captured_node, nil, nil
+          end
+          return iter(end_line) -- tail call: try next match
+        end
+
+        metadata = self:apply_directives(patterns.directives, pattern, captures, source)
+      else
+        metadata = {}
+      end
+
+      highest_cached_match_id = math.max(highest_cached_match_id, match_id)
```

Comment:
Is there a way to make this work without keeping track of the highest seen match id? Maybe by just checking if `match_cache[match_id]` exists? Just wondering

Reply:
Yes, it can be removed. Work correctly either way. 

I added this check since the cache might become a hash map at some point, and indexing it would become more expensive.

 Most of the time, captures returned by `next_capture()` are new and have match id higher than anything in the cache, so this check is cheaper.

It's about 1ms IIRC